feat(workspace-store): AsyncAPI support

hanspagel opened the pull request:
**Tasks**

- [ ] lots of conflicts, I’ll split up this PR into just the schemas/types and ff with the remaining stuff

**Problem**

Currently, we support OpenAPI documents, that's great. But what if someone wants to load an AsyncAPI document? 🤔 It seems so similar … so what if we add support for it?

**Solution**

This PR:
* adds AsyncAPI 3.0 schemas
* the workspace store returns `OpenApiDocument | AsyncApiDocument` types instead of just `OpenApiDocument`
* adds helpers to narrow down the type
* adds a little bit of type-narrowing with those helpers where necessary

**Out of scope**

* the navigation is wrong, that’s for another PR
* there’s basically no rendering for AsyncAPI documents, that’s for another PR

**Checklist**

I've gone through the following:

- [x] I've added an explanation _why_ this change is needed.
- [x] I've added a changeset (`pnpm changeset`).
- [x] I've added tests for the regression or new feature.
- [x] I've updated the documentation.

<!-- See the [contributing guide](https://github.com/scalar/scalar/blob/main/CONTRIBUTING.md) for more information. -->






<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Add AsyncAPI 3.0 support to the workspace store (schemas, navigation, server/client), gate OpenAPI-only UI/search paths, and update tests/docs.
> 
> - **Workspace Store**:
>   - **Schemas**: Introduce comprehensive `AsyncApiDocument` 3.0 schema set and navigation types; add `isOpenApiDocument`/`isAsyncApiDocument` type guards; support mixed `ApiDefinition`.
>   - **Navigation**: Add `createAsyncApiNavigation` with channel/operation/message traversal.
>   - **Server (SSR/Static)**: Externalize AsyncAPI operations; generate chunks for AsyncAPI; support mixed OpenAPI/AsyncAPI; update loaders and exports.
>   - **Client**: Accept AsyncAPI or OpenAPI, preserve original version (`x-original-asyncapi-version`), conditional navigation generation, config/override handling, and export/load logic.
>   - **Mutators**: Make request/server/security mutators tolerant; scope to OpenAPI where applicable.
> - **API Reference**:
>   - **Content/Models**: Guard OpenAPI-only features (`servers`, clients, models) via type checks.
>   - **Search**: Accept `OpenApiDocument | AsyncApiDocument`; build search index only for OpenAPI.
> - **Tests/Docs**:
>   - Add extensive AsyncAPI tests (client/server/navigation), update existing tests for new types, and expand README with AsyncAPI usage.
> - **Misc**:
>   - Update re-exports, add sample AsyncAPI source, and changeset for minor release.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 239f1c5dd126023867f94e955275028492b6faa8. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

hanspagel committed:
feat(asyncapi): add AsyncAPI schema definitions and validation tests

- Introduced AsyncAPI document schema and related types for validation.
- Added tests for validating AsyncAPI documents and ensuring correct identification between AsyncAPI and OpenAPI documents.
- Updated client and workspace schemas to support AsyncAPI integration.
- Implemented channel and operation schemas for AsyncAPI, enhancing the overall API structure.

hanspagel committed:
docs(changeset): feat: AsyncAPI support

hanspagel committed:
feat(asyncapi): enhance AsyncAPI schema definitions and introduce new components

- Refactored existing AsyncAPI schemas to improve type definitions and validation.
- Added new schemas for Binding, CorrelationId, Parameter, Reply, and MessageTrait.
- Updated ChannelItem and Operation schemas to utilize new types for better structure and clarity.
- Enhanced the index file to export new schemas, ensuring comprehensive access to all AsyncAPI components.
- Improved type safety and documentation throughout the AsyncAPI schema files.

hanspagel committed:
feat(asyncapi): implement AsyncAPI navigation generation and add tests

- Introduced `createAsyncApiNavigation` to generate navigation structure for AsyncAPI documents.
- Updated `createWorkspaceStore` to utilize the new navigation generation for AsyncAPI.
- Added tests for `traverseAsyncApiDocument` to ensure correct navigation entries for channels and operations.
- Enhanced type definitions for AsyncAPI operations and channels in the navigation schema.
- Created helper functions for traversing AsyncAPI documents, improving organization and clarity.

hanspagel committed:
feat(asyncapi): enhance workspace document handling and add tests for AsyncAPI integration

- Updated `createWorkspaceStore` to support storing and retrieving AsyncAPI documents alongside OpenAPI documents.
- Introduced tests to verify the correct storage and retrieval of both document types in the workspace.
- Enhanced type definitions to ensure proper type checking for AsyncAPI documents within the workspace.
- Implemented type guards to differentiate between OpenAPI and AsyncAPI documents, improving type safety.

hanspagel committed:
feat(asyncapi): update getDocument to support AsyncAPI documents

- Enhanced the `getDocument` function to return both OpenAPI and AsyncAPI documents, improving type safety and flexibility in document handling within the workspace.

hanspagel committed:
refactor(tests): enhance type safety for OpenAPI documents in tests

- Updated test cases to explicitly cast documents to OpenApiDocument, improving type safety and clarity.
- Modified helper functions and assertions to ensure compatibility with both OpenAPI and AsyncAPI document types.
- Enhanced the `requestMutators` function to accept both OpenApiDocument and AsyncApiDocument, ensuring proper handling of document types.

hanspagel committed:
docs(readme): enhance documentation for API support and mixed document handling

- Updated README to reflect support for both OpenAPI and AsyncAPI documents, improving clarity on the package's capabilities.
- Added examples for adding AsyncAPI documents to the store and retrieving mixed document types.
- Enhanced descriptions for server-side and client-side workspace stores to include AsyncAPI features.
- Included type-safe access examples for handling mixed document types in the workspace.

hanspagel committed:
feat(search): enhance document type support in Search components

- Updated SearchButton and SearchModal components to accept both OpenApiDocument and AsyncApiDocument types, improving flexibility in document handling.
- Modified useSearchIndex hook to accommodate mixed document types, ensuring proper search index creation based on the document type.

hanspagel committed:
refactor(tests): improve type safety for mock OpenAPI documents in tests

- Updated the `createMockDocument` function to return an OpenApiDocument type, enhancing type safety in test cases.
- Refactored import statements for better clarity and organization of OpenAPI document types.

hanspagel committed:
feat(tests): add OpenAPI document configuration to useSearchIndex tests

- Introduced OpenAPI document configuration in the useSearchIndex test cases to enhance testing scenarios.
- Updated test setup to include mock OpenAPI document details, improving the relevance and accuracy of search index tests.

hanspagel committed:
refactor(asyncapi): streamline schema definitions and enhance type safety

- Updated `createWorkspaceStore` to conditionally set version metadata based on document type, improving handling of OpenAPI and AsyncAPI documents.
- Refactored AsyncAPI schemas to utilize shared components, enhancing clarity and reducing redundancy.
- Improved type definitions for `ChannelsObject` and `OperationsObject`, ensuring better type safety and consistency across schemas.

hanspagel committed:
feat(tests): enhance useSearchIndex tests for AsyncAPI document handling

- Added comprehensive test cases for handling AsyncAPI documents in the useSearchIndex hook, ensuring it gracefully manages different document types.
- Introduced mock AsyncAPI and OpenAPI documents to validate search index generation and behavior under various scenarios.
- Improved assertions to confirm that search indices are only generated for OpenAPI documents, while AsyncAPI documents are handled correctly without generating indices.

hanspagel committed:
refactor(workspace-store): improve document type detection and upgrade logic

- Enhanced the `createWorkspaceStore` and `createServerWorkspaceStore` functions to detect document types before performing upgrades, ensuring only OpenAPI/Swagger documents are upgraded.
- Streamlined version metadata handling by storing original version information based on the detected document type, improving clarity and type safety.
- Refactored import statements for better organization and clarity in type definitions.

hanspagel committed:
test(workspace-store): improve type safety and null checks in client tests

- Added type imports and enhanced type safety for navigation and channel entries in client tests.
- Updated assertions to handle potential null values, ensuring robust checks for AsyncAPI and OpenAPI document structures.
- Refactored test cases to improve clarity and maintainability, particularly in handling document version metadata.

hanspagel committed:
feat(workspace-store): add AsyncAPI document processing and externalization

- Implemented support for AsyncAPI documents in the server workspace store, allowing for both SSR and static modes.
- Added `externalizeAsyncApiOperationReferences` function to generate references for AsyncAPI operations.
- Enhanced `createServerWorkspaceStore` to handle AsyncAPI documents alongside OpenAPI, including navigation structure generation.
- Updated tests to validate AsyncAPI document processing and ensure correct operation retrieval and externalization.

hanspagel committed:
refactor(workspace-store): enhance document type detection and type safety

- Refactored document type detection logic in `createWorkspaceStore` to utilize type guard functions for OpenAPI and AsyncAPI documents, improving clarity and maintainability.
- Updated type definitions to include `UnknownObject` for better type safety in document handling.
- Streamlined import statements for improved organization and clarity in type definitions.

DemonHa commented on packages/workspace-store/src/schemas/asyncapi/asyncapi-document.ts:
Can we prefix those schemas by the version like we are doing with the openapi schemas?

DemonHa commented on packages/workspace-store/src/schemas/workspace.ts:
```diff
@@ -14,13 +15,26 @@ import {
+export type ApiDefinition = OpenApiDocument | AsyncApiDocument
+
+// Type guards
+export function isOpenApiDocument(doc: ApiDefinition | UnknownObject): doc is OpenApiDocument {
```
maybe we can move those to the helpers

hanspagel replied:
ah! I thought I did already

hanspagel committed:
refactor(tests): simplify imports and fix mock collection usage in Operation tests

- Removed unused import of `collectionSchema` to streamline the test file.
- Updated the mock collection variable to ensure proper usage in the test cases, enhancing clarity and maintainability.

hanspagel committed:
chore: clean up

hanspagel committed:
refactor(workspace-store): reorganize type guards and improve imports

- Moved type guard functions for OpenAPI and AsyncAPI documents to a new helper file for better organization and clarity.
- Updated import statements across various files to utilize the new type guard location, enhancing maintainability.
- Added AsyncAPI document source to test data for improved testing coverage.

hanspagel committed:
refactor(workspace-store): migrate AsyncAPI schemas to v3.0 structure

- Updated import paths for AsyncAPI document types to reflect the new v3.0 schema organization.
- Removed deprecated AsyncAPI schema files and replaced them with new v3.0 schema definitions, enhancing clarity and maintainability.
- Ensured all references to AsyncAPI documents across the codebase are consistent with the new structure, improving type safety and organization.

hanspagel committed:
fix: broken imports

hanspagel committed:
fix: another broken import

hanspagel committed:
fix: more broken imports

hanspagel committed:
refactor(workspace-store): update AsyncAPI schema definitions and imports

- Refactored AsyncAPI schema files to export definitions directly, improving clarity and consistency.
- Updated import paths across various files to align with the new schema structure, enhancing maintainability.
- Ensured all AsyncAPI-related types and schemas are correctly referenced, improving type safety and organization.

hanspagel committed:
fix(workspace-store): handle undefined case in document version extraction

- Updated the logic in `createWorkspaceStore` to explicitly handle the case where both OpenAPI and Swagger properties are undefined, ensuring robustness in document version extraction.
- This change improves the reliability of the workspace store when processing various document types.

hanspagel committed:
refactor(workspace-store): enhance type guards for OpenAPI and Swagger documents

- Updated the type guard function to combine checks for OpenAPI and Swagger documents, improving clarity and maintainability.
- Refactored the `createWorkspaceStore` logic to utilize the new type guard, ensuring consistent document type detection.
- Added documentation comments for better understanding of the type guard functions.

hanspagel committed:
fix(api-reference): improve document version handling in Content component

- Enhanced the logic in the Content component to check if the document is an OpenAPI document before accessing its version property, preventing potential runtime errors.
- Updated the rendering condition for traversed operations to ensure it only displays when a valid OpenAPI document is present, improving robustness and user experience.

hanspagel committed:
fix(api-reference): update AsyncAPI import paths to v3.0 structure

- Corrected import paths for AsyncAPI document types in SearchButton, SearchModal, and useSearchIndex components to align with the new v3.0 schema organization.
- This change enhances clarity and maintainability by ensuring consistent references to AsyncAPI documents across the codebase.

hanspagel committed:
feat: improve asyncapi schemas

hanspagel committed:
feat: improve asyncapi schemas even more

hanspagel committed:
fix: asyncapi assets are not stored

hanspagel committed:
refactor: clean up

hanspagel committed:
fix: using old asycapi syntax

hanspagel committed:
feat(api-reference): add support for AsyncAPI document handling

- Introduced a check for AsyncAPI documents in the useWorkspaceStoreEvents hook to prepare for future support.
- Added TODO comments to indicate the need for implementing AsyncAPI functionality, enhancing clarity for upcoming development efforts.

hanspagel committed:
feat: add asyncapi security scheme schemas

hanspagel committed:
fix: disable asyncapi security scheme mutators

hanspagel committed:
feat: add Message Example Object

hanspagel committed:
fix: schemas fail to parse

hanspagel committed:
fix(workspace-store): update AsyncAPI operation actions for consistency

- Changed 'publish' to 'send' for user signup events.
- Updated 'subscribe' to 'receive' for user deletion events.
- Adjusted related test assertions to reflect these changes.

hanspagel committed:
fix(api-reference): ensure models are displayed for valid OpenAPI documents

- Updated the condition for rendering the Models component to check if the document is a valid OpenAPI document, preventing potential rendering issues.

hanspagel committed:
fix: TS error

hanspagel commented on packages/api-reference/src/components/Content/Content.vue:
```diff
@@ -142,7 +151,7 @@ const models = computed<TraversedDescription | undefined>(() => {
 
     <!-- Render traversed operations and webhooks -->
-    <div v-if="items.entries.length">
+    <div v-if="document && isOpenApiDocument(document) && items.entries.length">
```
skip the rendering for asyncapi

DemonHa commented on packages/workspace-store/src/schemas/asyncapi/v3.0/asyncapi-document.ts:
```diff
@@ -0,0 +1,323 @@
+    /** A map of Server Objects, which provide connectivity information to a target server. */
+    servers: Type.Optional(ServersObjectSchemaDefinition),
+    /** The available channels and operations for the API. */
+    channels: Type.Optional(ChannelsObjectSchemaDefinition),
```
same here and other places we are using the definition insted of the ref

hanspagel committed:
refactor: use refs

hanspagel committed:
refactor: use object syntax to match the naming in the specification

hanspagel committed:
feat: add dedicated object schemas for everything that’s in the specification

hanspagel committed:
feat: improve bindings

hanspagel committed:
refactor: use refs not definitions

hanspagel committed:
feat(workspace-store): enhance AsyncAPI message handling

- Introduced support for multi-format schemas in the Models component.
- Updated the rendering logic to include messages under channels in the AsyncAPI document.
- Added new type guards and utility functions for better message ID generation and schema resolution.
- Enhanced tests to cover new message handling scenarios, ensuring robust navigation structure for AsyncAPI documents.

hanspagel committed:
feat: hide client libraries for asyncapi

hanspagel committed:
refactor(api-reference): remove unused type MultiFormatSchemaObject from ClassicLayout component

- Simplified the schema prop type in ClassicLayout.vue by removing MultiFormatSchemaObject, ensuring it only uses SchemaObject.
- This change helps streamline the component's type definitions and improves clarity.

hanspagel committed:
fix(api-reference): correct title typo and enhance document validation check

- Fixed a typo in the title prop of the IntroductionCardItem component from "Authentication" to "Authpentication".
- Updated the condition for rendering the Models component to ensure it only displays for valid OpenAPI documents, improving the component's reliability.

hanspagel committed:
fix(api-reference): correct title typo in IntroductionCardItem component

- Fixed a typo in the title prop from "Authpentication" to "Authentication" in the Content.vue file, ensuring accurate display of the authentication title.

DemonHa commented on packages/workspace-store/src/navigation/helpers/traverse-asyncapi-document.ts:
```diff
@@ -0,0 +1,373 @@
+  //   | ((a: { action: string; channel: string }, b: { action: string; channel: string }) => number)
+
+  /** Function to generate unique IDs for markdown headings */
+  getHeadingId: (heading: { depth: number; value: string; slug?: string }) => string
```
We’re transitioning to a new interface for ID generation.
The new design uses a single function that accepts an object containing a type discriminator and its associated context.

DemonHa commented on packages/workspace-store/src/mutators/index.ts:
```diff
@@ -40,10 +41,20 @@ export function generateClientMutators(store: WorkspaceStore) {
       requestMutators: requestMutators(document),
-      securitySchemeMutators: securitySchemeMutators(document?.components?.securitySchemes),
+      securitySchemeMutators:
+        document && isOpenApiDocument(document)
```
instead of doing this we can create different mutators for async api and conditionally return mutators

DemonHa commented on packages/workspace-store/src/client.ts:
```diff
@@ -836,7 +870,7 @@ export const createWorkspaceStore = (workspaceProps?: WorkspaceProps): Workspace
           Object.entries(result.documents).map(([name, doc]) => [
             name,
-            createOverridesProxy(createMagicProxy(doc), result.overrides[name]),
+            createOverridesProxy(createMagicProxy(doc) as ApiDefinition, result.overrides[name]),
```
instead of casting we need to update the Inmem worksapce schema

Dustin4444 reviewed (commented):
Fixes #8288 Fixes #8348

Dustin4444 reviewed (commented):
#8288
